swarm/storage/pushsync/pusher_test.go

```diff
@@ -0,0 +1,414 @@
+}
+
+func (lb *loopBack) send(to []byte, topic string, msg []byte) error {
+	p := p2p.NewPeer(enode.ID{}, "", nil)
+	for _, handler := range lb.handlers[topic] {
+		if err := handler(msg, p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BaseAddr needed to implement PubSub interface
+func (lb *loopBack) BaseAddr() []byte {
+	return lb.addr
+}
+
+// testPushSyncIndex mocks localstore and provides subscription and setting synced status
+// it implements PushSyncIndex interface
+type testPushSyncIndex struct {
+	i, total int
+	tagIDs   []uint32 //
+	tags     *chunk.Tags
+	sent     *sync.Map // to store time of send for retry
+	synced   chan int  // to check if right amount of chunks
+}
+
+func newTestPushSyncIndex(chunkCnt int, tagIDs []uint32, tags *chunk.Tags, sent *sync.Map) *testPushSyncIndex {
+	return &testPushSyncIndex{
+		i:      0,
+		total:  chunkCnt,
+		tagIDs: tagIDs,
+		tags:   tags,
+		sent:   sent,
+		synced: make(chan int),
+	}
+}
+
+// SubscribePush allows iteration on the hashes and mocks the behaviour of localstore
+// push index
+// we keep track of an index so that each call to SubscribePush knows where to start
+// generating the new fake hashes
+// Before the new fake hashes it iterates over hashes not synced yet
+func (t *testPushSyncIndex) SubscribePush(context.Context) (<-chan storage.Chunk, func()) {
+	chunks := make(chan storage.Chunk)
+	tagCnt := len(t.tagIDs)
+	quit := make(chan struct{})
+	stop := func() { close(quit) }
+	go func() {
+		// feed fake chunks into the db, hashes encode the order so that
+		// it can be traced
+		feed := func(i int) bool {
+			// generate fake hashes that encode the chunk order
+			addr := make([]byte, 32)
+			binary.BigEndian.PutUint64(addr, uint64(i))
+			// remember when the chunk was put
+			// if sent again, dont modify the time
+			t.sent.Store(i, time.Now())
+			// increment stored count on tag
+			tagID := t.tagIDs[i%tagCnt]
+			if tag, _ := t.tags.Get(tagID); tag != nil {
+				tag.Inc(chunk.StateStored)
+			}
+			select {
+			// chunks have no data and belong to tag i%tagCount
+			case chunks <- storage.NewChunk(addr, nil).WithTag(tagID):
+				return true
+			case <-quit:
+				return false
+			}
+		}
+		// push the chunks already pushed but not yet synced
+		t.sent.Range(func(k, _ interface{}) bool {
+			log.Debug("resending", "cur", k)
+			return feed(k.(int))
+		})
+		// generate the new chunks from t.i
+		for t.i < t.total && feed(t.i) {
+			t.i++
+		}
+
+		log.Debug("sent all chunks", "total", t.total)
+		close(chunks)
+	}()
+	return chunks, stop
+}
+
+func (t *testPushSyncIndex) Set(ctx context.Context, _ chunk.ModeSet, addr storage.Address) error {
+	cur := int(binary.BigEndian.Uint64(addr[:8]))
+	t.sent.Delete(cur)
+	t.synced <- cur
+	log.Debug("set chunk synced", "cur", cur, "addr", addr)
+	return nil
+}
+
+var (
+	maxDelay       = 210 // max delay in millisecond
+	minDelay       = 1   // min delay in millisecond
+	retentionLimit = 200 // ~5% of msg lost
```

Comment:
what does this number measure?

Reply:
we artificially delay receipt responses, delays above this number mean never receive receipt, hence the comment ~5% of msg lost